File: riveredge-backend/src/apps/kuaizhizao/models/maintenance_plan.py
"""
维护保养计划模型模块

定义维护保养计划数据模型，支持多组织隔离。

Author: Luigi Lu
Date: 2025-01-15
"""

from tortoise import fields
from core.models.base import BaseModel


class MaintenancePlan(BaseModel):
    """
    维护保养计划模型
    
    用于管理设备维护保养计划，支持多组织隔离。
    
    注意：继承自 BaseModel，自动包含 uuid、tenant_id、created_at、updated_at 字段。
    
    Attributes:
        id: 主键ID（自增ID，内部使用）
        uuid: 业务ID（UUID，对外暴露，安全且唯一，继承自BaseModel）
        tenant_id: 组织ID（用于多组织数据隔离，继承自BaseModel）
        plan_no: 维护计划编号（组织内唯一）
        plan_name: 计划名称
        equipment_id: 设备ID（关联设备）
        equipment_uuid: 设备UUID
        equipment_name: 设备名称
        plan_type: 计划类型（预防性维护、定期维护、临时维护）
        maintenance_type: 维护类型（日常保养、小修、中修、大修）
        cycle_type: 周期类型（按时间、按运行时长、按使用次数）
        cycle_value: 周期值
        cycle_unit: 周期单位（天、小时、次）
        planned_start_date: 计划开始日期
        planned_end_date: 计划结束日期
        responsible_person_id: 负责人ID（用户ID）
        responsible_person_name: 负责人姓名
        status: 计划状态（草稿、已发布、执行中、已完成、已取消）
        remark: 备注
        created_at: 创建时间（继承自BaseModel）
        updated_at: 更新时间（继承自BaseModel）
        deleted_at: 删除时间（软删除）
    """
    
    class Meta:
        """
        模型元数据
        """
        table = "core_maintenance_plans"
        indexes = [
            ("tenant_id",),
            ("plan_no",),
            ("equipment_id",),
            ("equipment_uuid",),
            ("status",),
            ("planned_start_date",),
        ]
        unique_together = [("tenant_id", "plan_no")]
    
    # 主键（BaseModel 不包含 id 字段，需要自己定义）
    id = fields.IntField(pk=True, description="主键ID")
    
    # 基本信息
    plan_no = fields.CharField(max_length=100, description="维护计划编号（组织内唯一）")
    plan_name = fields.CharField(max_length=200, description="计划名称")
    
    # 关联设备
    equipment_id = fields.IntField(description="设备ID（关联设备）")
    equipment_uuid = fields.CharField(max_length=36, description="设备UUID")
    equipment_name = fields.CharField(max_length=200, description="设备名称")
    
    # 计划类型
    plan_type = fields.CharField(max_length=50, description="计划类型（预防性维护、定期维护、临时维护）")
    maintenance_type = fields.CharField(max_length=50, description="维护类型（日常保养、小修、中修、大修）")
    
    # 周期设置
    cycle_type = fields.CharField(max_length=50, description="周期类型（按时间、按运行时长、按使用次数）")
    cycle_value = fields.IntField(null=True, description="周期值")
    cycle_unit = fields.CharField(max_length=20, null=True, description="周期单位（天、小时、次）")
    
    # 计划时间
    planned_start_date = fields.DatetimeField(null=True, description="计划开始日期")
    planned_end_date = fields.DatetimeField(null=True, description="计划结束日期")
    
    # 负责人
    responsible_person_id = fields.IntField(null=True, description="负责人ID（用户ID）")
    responsible_person_name = fields.CharField(max_length=100, null=True, description="负责人姓名")
    
    # 状态
    status = fields.CharField(max_length=50, default="草稿", description="计划状态（草稿、已发布、执行中、已完成、已取消）")
    remark = fields.TextField(null=True, description="备注")
    
    # 软删除字段
    deleted_at = fields.DatetimeField(null=True, description="删除时间（软删除）")
    
    def __str__(self):
        """字符串表示"""
        return f"{self.plan_no} - {self.plan_name}"


class MaintenanceExecution(BaseModel):
    """
    维护执行记录模型
    
    用于管理设备维护执行记录，支持多组织隔离。
    
    注意：继承自 BaseModel，自动包含 uuid、tenant_id、created_at、updated_at 字段。
    
    Attributes:
        id: 主键ID（自增ID，内部使用）
        uuid: 业务ID（UUID，对外暴露，安全且唯一，继承自BaseModel）
        tenant_id: 组织ID（用于多组织数据隔离，继承自BaseModel）
        execution_no: 执行记录编号（组织内唯一）
        maintenance_plan_id: 维护计划ID（关联维护计划）
        maintenance_plan_uuid: 维护计划UUID
        equipment_id: 设备ID（关联设备）
        equipment_uuid: 设备UUID
        equipment_name: 设备名称
        execution_date: 执行日期
        executor_id: 执行人员ID（用户ID）
        executor_name: 执行人员姓名
        execution_content: 执行内容
        execution_result: 执行结果（正常、异常、待处理）
        maintenance_cost: 维护成本
        spare_parts_used: 使用备件（JSON格式）
        status: 记录状态（草稿、已确认、已验收）
        acceptance_person_id: 验收人员ID（用户ID）
        acceptance_person_name: 验收人员姓名
        acceptance_date: 验收日期
        acceptance_result: 验收结果（合格、不合格）
        remark: 备注
        created_at: 创建时间（继承自BaseModel）
        updated_at: 更新时间（继承自BaseModel）
        deleted_at: 删除时间（软删除）
    """
    
    class Meta:
        """
        模型元数据
        """
        table = "core_maintenance_executions"
        indexes = [
            ("tenant_id",),
            ("execution_no",),
            ("maintenance_plan_id",),
            ("equipment_id",),
            ("execution_date",),
            ("status",),
        ]
        unique_together = [("tenant_id", "execution_no")]
    
    # 主键（BaseModel 不包含 id 字段，需要自己定义）
    id = fields.IntField(pk=True, description="主键ID")
    
    # 基本信息
    execution_no = fields.CharField(max_length=100, description="执行记录编号（组织内唯一）")
    
    # 关联维护计划
    maintenance_plan_id = fields.IntField(null=True, description="维护计划ID（关联维护计划）")
    maintenance_plan_uuid = fields.CharField(max_length=36, null=True, description="维护计划UUID")
    
    # 关联设备
    equipment_id = fields.IntField(description="设备ID（关联设备）")
    equipment_uuid = fields.CharField(max_length=36, description="设备UUID")
    equipment_name = fields.CharField(max_length=200, description="设备名称")
    
    # 执行信息
    execution_date = fields.DatetimeField(description="执行日期")
    executor_id = fields.IntField(null=True, description="执行人员ID（用户ID）")
    executor_name = fields.CharField(max_length=100, null=True, description="执行人员姓名")
    execution_content = fields.TextField(null=True, description="执行内容")
    execution_result = fields.CharField(max_length=50, null=True, description="执行结果（正常、异常、待处理）")
    maintenance_cost = fields.DecimalField(max_digits=10, decimal_places=2, null=True, description="维护成本")
    spare_parts_used = fields.JSONField(null=True, description="使用备件（JSON格式）")
    
    # 验收信息
    status = fields.CharField(max_length=50, default="草稿", description="记录状态（草稿、已确认、已验收）")
    acceptance_person_id = fields.IntField(null=True, description="验收人员ID（用户ID）")
    acceptance_person_name = fields.CharField(max_length=100, null=True, description="验收人员姓名")
    acceptance_date = fields.DatetimeField(null=True, description="验收日期")
    acceptance_result = fields.CharField(max_length=50, null=True, description="验收结果（合格、不合格）")
    remark = fields.TextField(null=True, description="备注")
    
    # 软删除字段
    deleted_at = fields.DatetimeField(null=True, description="删除时间（软删除）")
    
    def __str__(self):
        """字符串表示"""
        return f"{self.execution_no} - {self.equipment_name}"

File: riveredge-frontend/src/services/maintenancePlan.ts
/**
 * 维护保养计划 API 服务
 * 
 * 提供维护保养计划相关的 API 接口
 * 注意：所有 API 自动过滤当前组织的维护计划
 */

import { apiRequest } from './api';

/**
 * 维护计划信息接口
 */
export interface MaintenancePlan {
  uuid: string;
  plan_no: string;
  plan_name: string;
  equipment_uuid: string;
  equipment_id: number;
  equipment_name: string;
  plan_type: string;
  maintenance_type: string;
  cycle_type: string;
  cycle_value?: number;
  cycle_unit?: string;
  planned_start_date?: string;
  planned_end_date?: string;
  responsible_person_id?: number;
  responsible_person_name?: string;
  status: string;
  remark?: string;
  tenant_id: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

/**
 * 维护执行记录信息接口
 */
export interface MaintenanceExecution {
  uuid: string;
  execution_no: string;
  maintenance_plan_id?: number;
  maintenance_plan_uuid?: string;
  equipment_id: number;
  equipment_uuid: string;
  equipment_name: string;
  execution_date: string;
  executor_id?: number;
  executor_name?: string;
  execution_content?: string;
  execution_result?: string;
  maintenance_cost?: number;
  spare_parts_used?: Record<string, any>;
  status: string;
  acceptance_person_id?: number;
  acceptance_person_name?: string;
  acceptance_date?: string;
  acceptance_result?: string;
  remark?: string;
  tenant_id: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

/**
 * 维护计划列表查询参数
 */
export interface MaintenancePlanListParams {
  skip?: number;
  limit?: number;
  equipment_uuid?: string;
  status?: string;
  plan_type?: string;
  search?: string;
}

/**
 * 维护计划列表响应数据
 */
export interface MaintenancePlanListResponse {
  items: MaintenancePlan[];
  total: number;
  skip: number;
  limit: number;
}

/**
 * 维护执行记录列表查询参数
 */
export interface MaintenanceExecutionListParams {
  skip?: number;
  limit?: number;
  equipment_uuid?: string;
  maintenance_plan_uuid?: string;
  status?: string;
  search?: string;
}

/**
 * 维护执行记录列表响应数据
 */
export interface MaintenanceExecutionListResponse {
  items: MaintenanceExecution[];
  total: number;
  skip: number;
  limit: number;
}

/**
 * 创建维护计划数据
 */
export interface CreateMaintenancePlanData {
  plan_no?: string;
  plan_name: string;
  equipment_uuid: string;
  plan_type: string;
  maintenance_type: string;
  cycle_type: string;
  cycle_value?: number;
  cycle_unit?: string;
  planned_start_date?: string;
  planned_end_date?: string;
  responsible_person_id?: number;
  responsible_person_name?: string;
  status?: string;
  remark?: string;
}

/**
 * 更新维护计划数据
 */
export interface UpdateMaintenancePlanData {
  plan_name?: string;
  plan_type?: string;
  maintenance_type?: string;
  cycle_type?: string;
  cycle_value?: number;
  cycle_unit?: string;
  planned_start_date?: string;
  planned_end_date?: string;
  responsible_person_id?: number;
  responsible_person_name?: string;
  status?: string;
  remark?: string;
}

/**
 * 创建维护执行记录数据
 */
export interface CreateMaintenanceExecutionData {
  execution_no?: string;
  maintenance_plan_uuid?: string;
  equipment_uuid: string;
  execution_date: string;
  executor_id?: number;
  executor_name?: string;
  execution_content?: string;
  execution_result?: string;
  maintenance_cost?: number;
  spare_parts_used?: Record<string, any>;
  status?: string;
  acceptance_person_id?: number;
  acceptance_person_name?: string;
  acceptance_date?: string;
  acceptance_result?: string;
  remark?: string;
}

/**
 * 更新维护执行记录数据
 */
export interface UpdateMaintenanceExecutionData {
  execution_date?: string;
  executor_id?: number;
  executor_name?: string;
  execution_content?: string;
  execution_result?: string;
  maintenance_cost?: number;
  spare_parts_used?: Record<string, any>;
  status?: string;
  acceptance_person_id?: number;
  acceptance_person_name?: string;
  acceptance_date?: string;
  acceptance_result?: string;
  remark?: string;
}

/**
 * 获取维护计划列表
 * 
 * 自动过滤当前组织的维护计划。
 * 
 * @param params - 查询参数
 * @returns 维护计划列表响应数据
 */
export async function getMaintenancePlanList(params?: MaintenancePlanListParams): Promise<MaintenancePlanListResponse> {
  return apiRequest<MaintenancePlanListResponse>('/apps/kuaizhizao/maintenance-plans', {
    params,
  });
}

/**
 * 获取维护计划详情
 * 
 * 自动验证组织权限：只能获取当前组织的维护计划。
 * 
 * @param maintenancePlanUuid - 维护计划 UUID
 * @returns 维护计划信息
 */
export async function getMaintenancePlanByUuid(maintenancePlanUuid: string): Promise<MaintenancePlan> {
  return apiRequest<MaintenancePlan>(`/apps/kuaizhizao/maintenance-plans/${maintenancePlanUuid}`);
}

/**
 * 创建维护计划
 * 
 * 自动设置当前组织的 tenant_id。
 * 
 * @param data - 维护计划创建数据
 * @returns 创建的维护计划信息
 */
export async function createMaintenancePlan(data: CreateMaintenancePlanData): Promise<MaintenancePlan> {
  return apiRequest<MaintenancePlan>('/apps/kuaizhizao/maintenance-plans', {
    method: 'POST',
    data,
  });
}

/**
 * 更新维护计划
 * 
 * 自动验证组织权限：只能更新当前组织的维护计划。
 * 
 * @param maintenancePlanUuid - 维护计划 UUID
 * @param data - 维护计划更新数据
 * @returns 更新后的维护计划信息
 */
export async function updateMaintenancePlan(maintenancePlanUuid: string, data: UpdateMaintenancePlanData): Promise<MaintenancePlan> {
  return apiRequest<MaintenancePlan>(`/apps/kuaizhizao/maintenance-plans/${maintenancePlanUuid}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除维护计划
 * 
 * 自动验证组织权限：只能删除当前组织的维护计划。
 * 
 * @param maintenancePlanUuid - 维护计划 UUID
 */
export async function deleteMaintenancePlan(maintenancePlanUuid: string): Promise<void> {
  return apiRequest<void>(`/apps/kuaizhizao/maintenance-plans/${maintenancePlanUuid}`, {
    method: 'DELETE',
  });
}

/**
 * 获取维护执行记录列表
 * 
 * 自动过滤当前组织的维护执行记录。
 * 
 * @param params - 查询参数
 * @returns 维护执行记录列表响应数据
 */
export async function getMaintenanceExecutionList(params?: MaintenanceExecutionListParams): Promise<MaintenanceExecutionListResponse> {
  return apiRequest<MaintenanceExecutionListResponse>('/apps/kuaizhizao/maintenance-plans/executions', {
    params,
  });
}

/**
 * 获取维护执行记录详情
 * 
 * 自动验证组织权限：只能获取当前组织的维护执行记录。
 * 
 * @param executionUuid - 维护执行记录 UUID
 * @returns 维护执行记录信息
 */
export async function getMaintenanceExecutionByUuid(executionUuid: string): Promise<MaintenanceExecution> {
  return apiRequest<MaintenanceExecution>(`/apps/kuaizhizao/maintenance-plans/executions/${executionUuid}`);
}

/**
 * 创建维护执行记录
 * 
 * 自动设置当前组织的 tenant_id。
 * 
 * @param data - 维护执行记录创建数据
 * @returns 创建的维护执行记录信息
 */
export async function createMaintenanceExecution(data: CreateMaintenanceExecutionData): Promise<MaintenanceExecution> {
  return apiRequest<MaintenanceExecution>('/apps/kuaizhizao/maintenance-plans/executions', {
    method: 'POST',
    data,
  });
}

/**
 * 更新维护执行记录
 * 
 * 自动验证组织权限：只能更新当前组织的维护执行记录。
 * 
 * @param executionUuid - 维护执行记录 UUID
 * @param data - 维护执行记录更新数据
 * @returns 更新后的维护执行记录信息
 */
export async function updateMaintenanceExecution(executionUuid: string, data: UpdateMaintenanceExecutionData): Promise<MaintenanceExecution> {
  return apiRequest<MaintenanceExecution>(`/apps/kuaizhizao/maintenance-plans/executions/${executionUuid}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除维护执行记录
 * 
 * 自动验证组织权限：只能删除当前组织的维护执行记录。
 * 
 * @param executionUuid - 维护执行记录 UUID
 */
export async function deleteMaintenanceExecution(executionUuid: string): Promise<void> {
  return apiRequest<void>(`/apps/kuaizhizao/maintenance-plans/executions/${executionUuid}`, {
    method: 'DELETE',
  });
}

